libraries/AP_Winch/AP_Winch.cpp

```diff
@@ -0,0 +1,124 @@
+    // @Description: Winch enable/disable
+    // @User: Standard
+    // @Values: 0:None,1:Servo with encoder
+    AP_GROUPINFO("_TYPE", 1, AP_Winch, config.type, 0),
+
+    // @Param: _RATE_MAX
+    // @DisplayName: Winch deploy or retract rate maximum
+    // @Description: Winch deploy or retract rate maximum.  Set to maximum rate with no load.
+    // @User: Standard
+    // @Range: 0 10
+    // @Units: m/s
+    AP_GROUPINFO("_RATE_MAX", 2, AP_Winch, config.rate_max, 1.0f),
+
+    // @Param: _LEN_CURR
+    // @DisplayName: Winch line current length
+    // @Description: Winch line current length
+    // @User: Standard
+    // @Range: 0 100
+    // @Units: m
+    AP_GROUPINFO("_LEN_CURR", 3, AP_Winch, config.length_curr, 0),
+
+    // @Param: _RATE_P
+    // @DisplayName: Winch control rate P gain
+    // @Description: Winch control rate P gain.  Converts rate error (in radians/sec) to pwm output (in the range -1 to +1)
+    // @Range: 0.100 2.000
+    // @User: Standard
+
+    // @Param: _RATE_I
+    // @DisplayName: Winch control I gain
+    // @Description: Winch control I gain.  Corrects long term error between the desired rate (in rad/s) and actual
+    // @Range: 0.000 2.000
+    // @User: Standard
+
+    // @Param: _RATE_IMAX
+    // @DisplayName: Winch control I gain maximum
+    // @Description: Winch control I gain maximum.  Constrains the output (range -1 to +1) that the I term will generate
+    // @Range: 0.000 1.000
+    // @User: Standard
+
+    // @Param: _RATE_D
+    // @DisplayName: Winch control D gain
+    // @Description: Winch control D gain.  Compensates for short-term change in desired rate vs actual
+    // @Range: 0.000 0.400
+    // @User: Standard
+
+    // @Param: _RATE_FILT
+    // @DisplayName: Winch control filter frequency
+    // @Description: Winch control input filter.  Lower values reduce noise but add delay.
+    // @Range: 1.000 100.000
+    // @Units: Hz
+    // @User: Standard
+    AP_SUBGROUPINFO(config.rate_pid, "_RATE_", 4, AP_Winch, AC_PID),
+
+    AP_GROUPEND
+};
+
+AP_Winch::AP_Winch()
+{
+    AP_Param::setup_object_defaults(this, var_info);
+}
+
+void AP_Winch::init(const AP_WheelEncoder *wheel_encoder)
+{
+    // return immediately if not enabled
+    if (!_enabled.get()) {
+        return;
+    }
+
+    switch(config.type.get()) {
+    case 0:
+        break;
+    case 1:
+        backend = new AP_Winch_Servo(config);
+        break;
+    default:
+        break;
+    }
+    if (backend != nullptr) {
+        backend->init(wheel_encoder);
+    }
+}
+
+// release specified length of cable (in meters)
+void AP_Winch::release_length(float length)
+{
+    config.length_desired = config.length_curr + length;
+    config.state = STATE_POSITION;
+}
+
+// deploy line at specified speed in m/s (+ve deploys line, -ve retracts line, 0 stops)
+void AP_Winch::set_desired_rate(float rate)
+{
+    config.rate_desired = constrain_float(rate, -config.rate_max, config.rate_max);
```

Comment:
This should use get_rate_max() because on any constrain if low > high then the constrain will fail to do anything, and if the user typed a negative rate max in it wouldn't be handled correctly. (An error condition for sure, but we don't stop them from putting in that kind of value in the param).

Reply:
sounds good, I'll change it to use get_rate_max()